torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1099,6 +1117,329 @@ def call_function(
         return super().call_function(tx, args, kwargs)
 
 
+class CannotConvertRangeToHigherOrder(TorchDynamoException):
+    pass
+
+
+class RangeHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    """
+    The general idea is that for a simple for loop like
+
+    ```
+    for i in range(10):
+        x = x + 1
+    ```
+
+    We can convert the loop body to a function and the loop itself
+    to a series of calls:
+
+    ```
+    def loop_body(local_0, local_1, local_...):
+        local_1 = local_1 + 1
+        return (local_0, local_1, local_...)
+
+    (i, x) = loop_body(0, x)
+    (i, x) = loop_body(1, x)
+    (i, x) = loop_body(2, x)
+    # ... 7 more times
+    ```
+
+    There are a few benefits to doing this, but mainly we bypass a ton
+    of stuff associated with tracing for loops. That itself provides some speedup.
+
+    Once we implement function compilation caching, or a higher order operator,
+    like jax.lax.fori_loop, for loops will compile in constant time, and perhaps even
+    better than what we currently have.
+    """
+
+    # Any of these nested means the control flow is too complex to be
+    # represented by a higher order op.
+    FORBIDDEN_OPCODES = {
+        "RETURN_VALUE",
+        "YIELD_VALUE",
+        "YIELD_FROM",
+        "GET_YIELD_FROM_ITER",
+        "SEND",
+        "MAKE_FUNCTION",
+        # This could be supported in the future.
+        "LOAD_DEREF",
+        # Exceptions
+        "RAISE_VARARGS",
```

Comment:
Exceptions in 3.11+ are trickier to deal with because bytecodes denoting the start and end of a try block are gone and instead are replaced by an exception table.

3.12 in particular is really nasty because the except block bytecode is placed after the END_FOR.

Summary: we need to pay attention to the exception table for 3.11+.

Reply:
Are you also forbidding with blocks too? Because they seem to contain `POP_EXCEPT` in 3.9+. But in 3.8, with blocks don't contain any of `RAISE_VARARGS, SETUP_FINALLY, POP_EXCEPT`.

I still think in 3.11+, you will need to look at the exception table. If with blocks are forbidden, then you just need forbid instructions that have an exception table entry. If with blocks are not forbidden, then you will have to additionally determine if the exception table entry corresponds to a with block or not, by looking for the `BEFORE_WITH` instruction.

Reply:
I'm not but I think we can start simple for now. I've banned anything containing an exception table entry for now.